code/controllers/subsystems/typing.dm

```diff
@@ -0,0 +1,208 @@
+	if(timeout != entry[INDEX_TIMEOUT]) // We're stale. Touch nothing.
+		return
+	entry[INDEX_INFLIGHT] = FALSE
+	if(QDELETED(client) || !isliving(client.mob))
+		return
+	entry[INDEX_INPUT_STATE] = match_verbs.Find(content) != 0
+	update_indicator(client, entry)
+
+/// Attempt to update the mob's typing state and indicator according to new state.
+/datum/controller/subsystem/typing/proc/update_indicator(client/client, list/entry)
+	PRIVATE_PROC(TRUE)
+	var/mob/target = client.mob
+	var/display = target.stat == CONSCIOUS && entry[INDEX_PREFERENCE] && (entry[INDEX_INPUT_STATE] || entry[INDEX_VERB_STATE]) && !target.is_cloaked() && isturf(target.loc)
+	if(display == target.is_typing)
+		return
+	if(display)
+		if(!target.typing_indicator)
+			target.typing_indicator = new(target)
+		target.is_typing = TRUE
+		target.typing_indicator.show_typing_indicator()
+	else if(target.typing_indicator)
+		target.is_typing = FALSE
+		target.typing_indicator.hide_typing_indicator()
+
+/*
+Typing indicators, when a mob uses the F3/F4 keys to bring the say/emote input boxes up this little buddy is
+made and follows them around until they are done (or something bad happens), helps tell nearby people that 'hey!
+I IS TYPIN'!'
+Updated 09/10/2022 to include chatbar using Spookerton's SStyping system from Polaris.
+*/
+
+/atom/movable/overlay/typing_indicator
+	expected_master_type = /atom/movable
+	follow_proc = /atom/movable/proc/move_to_turf_or_null
+	icon = 'icons/mob/talk.dmi'
+	icon_state = "typing"
+	mouse_opacity = 0
+	vis_flags = VIS_INHERIT_ID
+
+/atom/movable/overlay/typing_indicator/Destroy()
+	if(ismovable(master))
+		var/atom/movable/AM = master
+		remove_vis_contents(AM, src)
+		if(ismob(master))
+			var/mob/owner = master
+			if(owner.typing_indicator == src)
+				owner.typing_indicator = null
+	return ..()
+
+/atom/movable/overlay/typing_indicator/proc/hide_typing_indicator()
+	set waitfor = FALSE
+	if(ismob(master))
+		var/mob/owner = master
+		if(owner.is_typing)
+			return
+	animate(src, alpha = 0, time = 0.5 SECONDS, easing = EASE_IN)
+	sleep(0.5 SECONDS)
+	set_invisibility(INVISIBILITY_MAXIMUM)
+	if(ismovable(master))
+		var/atom/movable/owner = master
+		remove_vis_contents(owner, src)
+
+/atom/movable/overlay/typing_indicator/proc/show_typing_indicator()
+
+	// Make it visible after being hidden.
+	set_invisibility(0)
+
+	// Animate it popping up from nowhere.
+	var/matrix/M = matrix()
+	M.Scale(0, 0)
+	transform = M
+	alpha = 0
+	animate(src, transform = 0, alpha = 255, time = 0.2 SECONDS, easing = EASE_IN)
+
+	if(ismob(master))
+		var/mob/owner = master
+		var/speech_state_modifier = owner.get_speech_bubble_state_modifier()
+		if(speech_state_modifier)
+			icon_state = "[speech_state_modifier]_typing"
+		else
+			icon_state = initial(icon_state)
+
+	if(ismovable(master))
+		var/atom/movable/owner = master
+		add_vis_contents(owner, src)
+
+/atom/movable/overlay/typing_indicator/SetInitLoc()
+	forceMove(get_turf(master))
```

Comment:
Doesn't this make it visible twice, once through vis contents and the second time just due to being at this loc? I would have assumed it should be positioned in null.